### PR TITLE
[Relanding] Implemented torch.linalg.multi_dot

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55064 Replace torch.chain_matmul calls to torch.linalg.multi_dot
* #53453 Deprecate torch.chain_matmul in favor of torch.linalg.multi_dot
* **#52859 [Relanding] Implemented torch.linalg.multi_dot**

This reverts commit 92a4ee1cf6092dd941591f80885eb7fef5b2c0d8.

Added support for bfloat16 for CUDA 11 and removed fast-path for empty input tensors that was affecting autograd graph.

Differential Revision: [D27402390](https://our.internmc.facebook.com/intern/diff/D27402390)

### CODE REVIEW

**Review by malfet — APPROVED** [review]

File: test/test_linalg.py [hunk 1]
```diff
@@ -3226,6 +3226,72 @@ def test_old_matrix_rank(self, device, dtype):
...
+        check([3, 2], [2, 2], [2, 3], [3, 4], discontiguous=True)
+
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.float)
+    def test_multi_dot_errors(self, device, dtype):
+        def check(tensors, out, msg):
+            with self.assertRaisesRegex(RuntimeError, msg):
+                torch.linalg.multi_dot(tensors, out=out)
+
+        a = torch.randn(2, device=device, dtype=dtype)
```
Comment (mruberry) [inline test/test_linalg.py:3274]: Not a PR blocking issue, but prefer make_tensor().

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -1811,6 +1832,26 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
            decorators=[_wrap_maybe_warns("floor_divide is deprecated, and will be removed")],
            supports_autograd=False,
            ),
+    OpInfo('linalg.multi_dot',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1835]: Not for this PR, but we should make a plan for the existing chain_matmul() tests like 

https://github.com/pytorch/pytorch/blob/f5617b0932c711ff4701904e9462a29fcb020d70/test/test_autograd.py#L2972

**Review by mruberry — APPROVED** [review]
This is really cool; nice work!

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -389,6 +389,25 @@ def sample_inputs_tensor_split(op_info, device, dtype, requires_grad):
                         args=(torch.tensor([1, 2, 3]),),
                         kwargs=dict(dim=1)),)
 
+def sample_inputs_linalg_multi_dot(op_info, device, dtype, requires_grad):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:392]: How long do these tests take? Replacing "S" with 3 would probably decrease the time taken.

**Review by mruberry — APPROVED** [review]
This PR looks excellent to me. Some remaining mypy issues need to be addressed. @IvanYashchuk, would you like to take a look at this, too?

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py:392]: Overall 10s, I tired replacing S with 3 but there's almost no difference.

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:392]: 10s is OK for now.

**Comment by vadimkantorov** [comment]
Does multi_dot support batched products i.e. processing a list of matrices of shapes BxP1xP2, BxP2xP3, BxP3xP4?

Also may be interesting to have a matrix product reduction method, e.g. for a tensor of BxWxPxP and dim it would be equivalent of batched `multi_dot(*input.unbind(dim))` (but maybe having more efficient computation because of known contiguity / layout)

**Comment by heitorschueroff** [comment]
> Does multi_dot support batched products i.e. processing a list of matrices of shapes BxP1xP2, BxP2xP3, BxP3xP4?
> 
> Also may be interesting to have a matrix product reduction method, e.g. for a tensor of BxWxPxP and dim it would be equivalent of batched `multi_dot(*input.unbind(dim))` (but maybe having more efficient computation because of known contiguity / layout)

It does not currently support broadcasting which is in line with NumPy. However, we could consider adding broadcast in a future PR.

**Comment by vadimkantorov** [comment]
I am not even talking of broadcasting, just regular batching...

**Comment by heitorschueroff** [comment]
> I am not even talking of broadcasting, just regular batching...

My mistake, I meant to say batching. It does not support batching.
